src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs

```diff
@@ -266,50 +278,81 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
-        /// The number 37 was chosen because that number allows for a perfect hash of the set of
-        /// "known strings" (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE, HTTP/1.0 and HTTP/1.1, where strings
-        /// with less than 8 characters have 0s appended to their ends to fill for the missing bytes).
+        /// in that format, it can be checked agains the known method. 
+        /// The Known Methods (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE) are all less then 8 bytes 
+        /// and will be compared with the required space. A mask is used if the Known method is less then 8 bytes.
+        /// To optimize performance the GET method will be checked first
         /// </remarks>
         /// <param name="begin">The iterator from which to start the known string lookup.</param>
-        /// <param name="end">The iterator pointing to the end of the input string.</param>
-        /// <param name="knownString">A reference to a pre-allocated known string, if the input matches any.</param>
+        /// <param name="scan">If we found a valid method, the scan will be updated to new position</param>
+        /// <param name="knownMethod">A reference to a pre-allocated known string, if the input matches any.</param>
         /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
-        public static bool GetKnownString(this MemoryPoolIterator2 begin, MemoryPoolIterator2 end, out string knownString)
+        public static bool GetKnownMethod(this MemoryPoolIterator2 begin, ref MemoryPoolIterator2 scan, out string knownMethod)
         {
-            knownString = null;
+            knownMethod = null;
+            var value = begin.PeekLong();
 
-            // This optimization only works on little endian environments (for now).
-            if (!BitConverter.IsLittleEndian)
+            if ((value & _mask4Chars) == _httpGetMethodLong)
             {
-                return false;
+                knownMethod = HttpGetMethod;
+                scan.Skip(4);
+                return true;
             }
-
-            var inputLength = begin.GetLength(end);
-
-            if (inputLength > sizeof(long))
+            foreach (var x in _knownMethods)
             {
-                return false;
+                if ((value & x.Item1) == x.Item2)
+                {
+                    knownMethod = x.Item3;
+                    scan.Skip(knownMethod.Length + 1);
+                    return true;
+                }
             }
 
-            var inputLong = begin.PeekLong();
+            return false;
+        }
 
-            if (inputLong == -1)
+        /// <summary>
+        /// Checks that up to 8 bytes between <paramref name="begin"/> and <paramref name="end"/> correspond to a known HTTP string.
+        /// </summary>
+        /// <remarks>
+        /// A "known HTTP version" Is a version defined in the HTTP/1.1 (HTTP/1.0 or HTTP/1.1).
```

Comment:
I'd replace this sentence with `A "known HTTP version" is either HTTP/1.0 or HTTP/1.1.`